<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Mi-empresa" data-toc-modified-id="Mi-empresa-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Mi empresa</a></span></li><li><span><a href="#API-Foursquare." data-toc-modified-id="API-Foursquare.-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>API Foursquare.</a></span><ul class="toc-item"><li><span><a href="#Primera-llamada:-Aeropuerto" data-toc-modified-id="Primera-llamada:-Aeropuerto-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Primera llamada: Aeropuerto</a></span><ul class="toc-item"><li><span><a href="#El-aeropuerto-más-cercano-está-a-19,5km-:" data-toc-modified-id="El-aeropuerto-más-cercano-está-a-19,5km-:-2.1.1"><span class="toc-item-num">2.1.1&nbsp;&nbsp;</span>El aeropuerto más cercano está a 19,5km :</a></span></li></ul></li><li><span><a href="#Segunda-llamada:-Nursery-School" data-toc-modified-id="Segunda-llamada:-Nursery-School-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Segunda llamada: Nursery School</a></span><ul class="toc-item"><li><span><a href="#¿Cuál-de-las-Nurseries-está-más-cerca-de-mi-empresa?" data-toc-modified-id="¿Cuál-de-las-Nurseries-está-más-cerca-de-mi-empresa?-2.2.1"><span class="toc-item-num">2.2.1&nbsp;&nbsp;</span>¿Cuál de las Nurseries está más cerca de mi empresa?</a></span></li><li><span><a href="#La-Nursery-más-cercana-está-a--2,5km:" data-toc-modified-id="La-Nursery-más-cercana-está-a--2,5km:-2.2.2"><span class="toc-item-num">2.2.2&nbsp;&nbsp;</span>La Nursery más cercana está a  2,5km:</a></span></li></ul></li><li><span><a href="#Tercera-llamada:-Nightclubs" data-toc-modified-id="Tercera-llamada:-Nightclubs-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Tercera llamada: Nightclubs</a></span><ul class="toc-item"><li><span><a href="#¿Dónde-está-el-Nightclub-más-cercano-a-mi-empresa?" data-toc-modified-id="¿Dónde-está-el-Nightclub-más-cercano-a-mi-empresa?-2.3.1"><span class="toc-item-num">2.3.1&nbsp;&nbsp;</span>¿Dónde está el Nightclub más cercano a mi empresa?</a></span></li><li><span><a href="#El-nightclub-más-cercano-está-a-1,7km-:" data-toc-modified-id="El-nightclub-más-cercano-está-a-1,7km-:-2.3.2"><span class="toc-item-num">2.3.2&nbsp;&nbsp;</span>El nightclub más cercano está a 1,7km :</a></span></li></ul></li><li><span><a href="#Cuarta-llamada:-Starbucks" data-toc-modified-id="Cuarta-llamada:-Starbucks-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Cuarta llamada: Starbucks</a></span><ul class="toc-item"><li><span><a href="#El-Starbucks-más-cercano-está-a-500m:" data-toc-modified-id="El-Starbucks-más-cercano-está-a-500m:-2.4.1"><span class="toc-item-num">2.4.1&nbsp;&nbsp;</span>El Starbucks más cercano está a 500m:</a></span></li></ul></li><li><span><a href="#Quinta-llamada:-Restaurante-Vegano" data-toc-modified-id="Quinta-llamada:-Restaurante-Vegano-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Quinta llamada: Restaurante Vegano</a></span><ul class="toc-item"><li><span><a href="#¿Cuál-es-el-restaurante-vegano-más-cercano-a-mi-empresa?" data-toc-modified-id="¿Cuál-es-el-restaurante-vegano-más-cercano-a-mi-empresa?-2.5.1"><span class="toc-item-num">2.5.1&nbsp;&nbsp;</span>¿Cuál es el restaurante vegano más cercano a mi empresa?</a></span></li><li><span><a href="#El-Restaurante-vegano-más-cercano-está-a-1.5km:" data-toc-modified-id="El-Restaurante-vegano-más-cercano-está-a-1.5km:-2.5.2"><span class="toc-item-num">2.5.2&nbsp;&nbsp;</span>El Restaurante vegano más cercano está a 1.5km:</a></span></li></ul></li></ul></li></ul></div>

In [1]:
import requests
import json
from dotenv import load_dotenv
import os
import pandas as pd
from functools import reduce
import operator
import geopandas as gpd
from pymongo import MongoClient,GEOSPHERE
import shapely.geometry
import geopandas as gpd
from cartoframes.viz import Map, Layer, popup_element
import sys
sys.path.append("../")
import src.funcionesok as fun

#  Mi empresa
Teniendo en cuenta mi empresa hago llamadas a la API.

In [2]:
company = "21 Mikonkatu, Helsinki FI 00100"

In [3]:
suomi = pd.read_csv('../data/dataOK.csv', encoding = 'ISO - 8859-1')
suomi

city country_code   latitude  longitude   name category_code  \
0  Helsinki          FIN  60.171821  24.945167  Apaja   games_video   

   number_of_employees  founded_year  
0                   44          2001

In [4]:
fun.geocode(company)

{'type': 'Point', 'coordinates': [60.17212, 24.94519]}

In [5]:
load_dotenv()

True

In [6]:
tok1 = os.getenv("tok1")
tok2 = os.getenv("tok2")

# API Foursquare.

Una vez seleccionada la oficina perfecta que cumple nuestras necesidades vamos a comprobar mediante la API de Foursquare los siguientes puntos;
- Los Account managers viajan mucho. Voy a busar un aeropuerto en un radio de 20kms.
- El 30% de los empleados de la compañía tiene al menos un niño. Voy a buscar guarderías en un radio de 5km.
- La edad de los trabajadores oscila entre 25 y 40 años. Voy a buscar discotecas en un radio de 10km.
- A los ejecutivos les gusta mucho la cadena de cafés "Starbucks". Voy a buscar Starbucks en un radio de 500m.
- La CEO de la compañía es vegana. Voy a buscar restaurantes veganos en un radio de 2km.

## Primera llamada: Aeropuerto

In [7]:
fun.geo_api()

¿Qué quieres buscar en Foursquare? 
Airport
¿Cuántas búsquedas quieres que muestre? 
1
¿En qué radio?(medida = metros) 
20000


In [ ]:
emp1:

In [11]:
url1= 'https://api.foursquare.com/v2/venues/explore'

In [12]:
params = {"client_id" : tok1,
              "client_secret" : tok2,
              "v": "20180323",
              "ll": f"60.17212, 24.94519",
              "query": "Airport",
              "limit": 1,
              "radius": 20000,
             }

In [13]:
resp = requests.get(url=url1, params=params)
airdata = json.loads(resp.text)


In [14]:
airdata.keys()

dict_keys(['meta', 'response'])

In [15]:
decoding_data = airdata.get("response")

In [16]:
decoded = decoding_data.get("groups")[0]

In [17]:
airp = decoded.get("items")

In [18]:
mapa_nombre = ["venue","name"]
m_latitud = ["venue","location","lat"]
m_longitud = ["venue","location","lng"]

In [19]:
def getFromDict(diccionario,mapa):
    return reduce (operator.getitem,mapa,diccionario)

In [20]:
print(getFromDict(airp[0],mapa_nombre))

Helsinki Airport (HEL) (Helsinki-Vantaan lentoasema)


In [21]:
print(getFromDict(airp[0],m_latitud))

60.31912855684579


In [22]:
print(getFromDict(airp[0],m_longitud))

24.968297481536865


In [23]:
aeropuerto = []
for dic in airp:
    ai = {}
    ai["name"] = getFromDict(dic,mapa_nombre)
    ai["latitud"] = getFromDict(dic,m_latitud)
    ai["longitud"] = getFromDict(dic,m_longitud)
    aeropuerto.append(ai)

In [24]:
aeropuerto[0]

{'name': 'Helsinki Airport (HEL) (Helsinki-Vantaan lentoasema)',
 'latitud': 60.31912855684579,
 'longitud': 24.968297481536865}

In [25]:
dataframefinal = pd.DataFrame(aeropuerto)
dataframefinal.head()

name    latitud   longitud
0  Helsinki Airport (HEL) (Helsinki-Vantaan lento...  60.319129  24.968297

In [26]:
gdf = gpd.GeoDataFrame(dataframefinal, geometry = gpd.points_from_xy(dataframefinal.longitud,dataframefinal.latitud))
gdf.head()

name    latitud   longitud  \
0  Helsinki Airport (HEL) (Helsinki-Vantaan lento...  60.319129  24.968297   

                    geometry  
0  POINT (24.96830 60.31913)

In [ ]:
#Creo un mapa con el punto selecccionado

In [ ]:
Map(Layer(gdf, popup_hover = [popup_element("name","Airport")]))

### El aeropuerto más cercano está a 19,5km : 
    Helsinki Airport (HEL) (Helsinki-Vantaan lentoasema)
    latitud: 60.319129
    longitud: 24.968297

In [49]:
fun.geo_api()

¿Qué quieres buscar en Foursquare? 
Nursery School
¿Cuántas búsquedas quieres que muestre? 
10
¿En qué radio?(medida = metros) 
5000


## Segunda llamada: Nursery School

In [29]:
params = {"client_id" : tok1,
              "client_secret" : tok2,
              "v": "20180323",
              "ll": f"60.17212, 24.94519",
              "query": "Nursery School",
              "limit": 10,
              "radius": 5000
             }

In [30]:
resp = requests.get(url=url1, params=params)
nurdata = json.loads(resp.text)

In [31]:
nurdata.keys()

dict_keys(['meta', 'response'])

In [32]:
decoding_data = nurdata.get("response")

In [33]:
decoded = decoding_data.get("groups")[0]

In [34]:
nur = decoded.get("items")


In [35]:
mapan_nombre = ["venue","name"]
mn_latitud = ["venue","location","lat"]
mn_longitud = ["venue","location","lng"]

In [36]:
def getFromDict(diccionario,mapa):
    return reduce (operator.getitem,mapa,diccionario)

In [37]:
print(getFromDict(nur[0],mapan_nombre))
print(getFromDict(nur[0],mn_latitud))
print(getFromDict(nur[0],mn_longitud))

Daghemmet B. E. Von Schantz
60.158121
24.936681


In [38]:
nurseries = []
for dic in nur:
    nr = {}
    nr["name"] = getFromDict(dic,mapa_nombre)
    nr["latitud"] = getFromDict(dic,m_latitud)
    nr["longitud"] = getFromDict(dic,m_longitud)
    nurseries.append(nr)

In [39]:
nurseries[0]

{'name': 'Daghemmet B. E. Von Schantz',
 'latitud': 60.158121,
 'longitud': 24.936681}

In [40]:
df_nur = pd.DataFrame(nurseries)
df_nur.head()

name    latitud   longitud
0     Daghemmet B. E. Von Schantz  60.158121  24.936681
1  lauttasaari english playschool  60.158857  24.877429
2             Päiväkoti Kesäheinä  60.202455  24.951274
3               Päiväkoti Marjala  60.209475  24.964812

In [41]:
gdf2 = gpd.GeoDataFrame(df_nur, geometry = gpd.points_from_xy(df_nur.longitud,df_nur.latitud))
gdf2.head()

name    latitud   longitud  \
0     Daghemmet B. E. Von Schantz  60.158121  24.936681   
1  lauttasaari english playschool  60.158857  24.877429   
2             Päiväkoti Kesäheinä  60.202455  24.951274   
3               Päiväkoti Marjala  60.209475  24.964812   

                    geometry  
0  POINT (24.93668 60.15812)  
1  POINT (24.87743 60.15886)  
2  POINT (24.95127 60.20246)  
3  POINT (24.96481 60.20948)

In [42]:
Map(Layer(gdf2, popup_hover = [popup_element("name","Nursery School")]))

### ¿Cuál de las Nurseries está más cerca de mi empresa?

In [48]:
gdf2['geometry']=gdf2['geometry'].apply(lambda x:shapely.geometry.mapping(x))
gdf2

AttributeError: 'dict' object has no attribute '__geo_interface__'

In [44]:
client = MongoClient()
db = client.ironhack
collection = db.nurseries
collection.create_index ([("geometry", GEOSPHERE)])

'geometry_2dsphere'

In [45]:
data = gdf2.to_dict(orient='records')
collection.insert_many(data)

In [46]:
data

[{'name': 'Daghemmet B. E. Von Schantz',
  'latitud': 60.158121,
  'longitud': 24.936681,
  'geometry': {'type': 'Point', 'coordinates': (24.936681, 60.158121)},
  '_id': ObjectId('6021b00806255645aa2c526f')},
 {'name': 'lauttasaari english playschool',
  'latitud': 60.158857316461074,
  'longitud': 24.877429136253458,
  'geometry': {'type': 'Point',
   'coordinates': (24.877429136253458, 60.158857316461074)},
  '_id': ObjectId('6021b00806255645aa2c5270')},
 {'name': 'Päiväkoti Kesäheinä',
  'latitud': 60.20245514603548,
  'longitud': 24.95127417000266,
  'geometry': {'type': 'Point',
   'coordinates': (24.95127417000266, 60.20245514603548)},
  '_id': ObjectId('6021b00806255645aa2c5271')},
 {'name': 'Päiväkoti Marjala',
  'latitud': 60.20947535503745,
  'longitud': 24.964811971036774,
  'geometry': {'type': 'Point',
   'coordinates': (24.964811971036774, 60.20947535503745)},
  '_id': ObjectId('6021b00806255645aa2c5272')}]

In [47]:
cercano2 = db.nurseries.find({"geometry":{"$near":{"type":"Point", "coordinates":[60.171821,24.945167]}}})
list(cercano2)

[{'_id': ObjectId('60212ef3235f9908d38c87d0'),
  'name': 'Daghemmet B. E. Von Schantz',
  'latitud': 60.158121,
  'longitud': 24.936681,
  'geometry': {'type': 'Point', 'coordinates': [24.936681, 60.158121]}},
 {'_id': ObjectId('6021b00806255645aa2c526f'),
  'name': 'Daghemmet B. E. Von Schantz',
  'latitud': 60.158121,
  'longitud': 24.936681,
  'geometry': {'type': 'Point', 'coordinates': [24.936681, 60.158121]}},
 {'_id': ObjectId('602164a3bda0fd886bdf4fe1'),
  'name': 'Daghemmet B. E. Von Schantz',
  'latitud': 60.158121,
  'longitud': 24.936681,
  'geometry': {'type': 'Point', 'coordinates': [24.936681, 60.158121]}},
 {'_id': ObjectId('6021a68b0a89d8e223fafa20'),
  'name': 'Daghemmet B. E. Von Schantz',
  'latitud': 60.158121,
  'longitud': 24.936681,
  'geometry': {'type': 'Point', 'coordinates': [24.936681, 60.158121]}},
 {'_id': ObjectId('6021b00806255645aa2c5272'),
  'name': 'Päiväkoti Marjala',
  'latitud': 60.20947535503745,
  'longitud': 24.964811971036774,
  'geometry': {'

### La Nursery más cercana está a  2,5km: 
    Daghemmet B. E. Von Schantz
    latitud': 60.158121
    longitud': 24.936681

## Tercera llamada: Nightclubs

In [ ]:
params = {"client_id" : tok1,
              "client_secret" : tok2,
              "v": "20180323",
              "ll": f"60.17212, 24.94519",
              "query": "Nightlife Spot",
              "limit": 25,
              "radius": 7500 
         }
resp = requests.get(url=url1, params=params)
nidata = json.loads(resp.text)

In [ ]:
nidata.keys()

In [ ]:
decoding_data = nidata.get("response")

In [ ]:
decoded = decoding_data.get("groups")[0]


In [ ]:
nic = decoded.get("items")


In [ ]:
mapani_nombre = ["venue","name"]
mapani_latitud = ["venue","location","lat"]
mapani_longitud = ["venue","location","lng"]

In [ ]:
print(getFromDict(nic[0],mapa_nombre))
print(getFromDict(nic[0],m_latitud))
print(getFromDict(nic[0],m_longitud))

In [ ]:
nightclubs = []
for dic in nic:
    nc = {}
    nc["name"] = getFromDict(dic,mapa_nombre)
    nc["latitud"] = getFromDict(dic,m_latitud)
    nc["longitud"] = getFromDict(dic,m_longitud)
    nightclubs.append(nc)

In [ ]:
nightclubs[0]

In [ ]:
df_nc = pd.DataFrame(nightclubs)
df_nc.head()

In [ ]:
gdf3 = gpd.GeoDataFrame(df_nc, geometry = gpd.points_from_xy(df_nc.longitud,df_nc.latitud))
gdf3.head()

In [ ]:
Map(Layer(gdf3, popup_hover = [popup_element("name","Night Club")]))

### ¿Dónde está el Nightclub más cercano a mi empresa?

In [ ]:
gdf3['geometry']=gdf3['geometry'].apply(lambda x:shapely.geometry.mapping(x))
gdf3

In [ ]:
db = client.ironhack
collection = db.nightclubs
collection.create_index ([("geometry", GEOSPHERE)])

In [ ]:
data = gdf3.to_dict(orient='records')
collection.insert_many(data)

In [ ]:
data[:5]

In [ ]:
cercano3 = db.nightclubs.find({"geometry":{"$near":{"type":"Point", "coordinates":[60.171821,24.945167]}}})
list(cercano3)[:5]

### El nightclub más cercano está a 1,7km : 
    Bar Haven
    latitud': 60.16654471941996
    longitud': 24.9513536704468

## Cuarta llamada: Starbucks

In [ ]:
params = {"client_id" : tok1,
              "client_secret" : tok2,
              "v": "20180323",
              "ll": f"60.17212, 24.94519",
              "query": "Starbucks",
              "limit": 10,
              "radius": 500
         }
resp = requests.get(url=url1, params=params)
strdata = json.loads(resp.text)


In [ ]:
strdata.keys()

In [ ]:
decoding_data = strdata.get("response")

In [ ]:
decoded = decoding_data.get("groups")[0]

In [ ]:
star = decoded.get("items")

In [ ]:
mapastar_nombre = ["venue","name"]
mapastar_latitud = ["venue","location","lat"]
mapastar_longitud = ["venue","location","lng"]

In [ ]:
print(getFromDict(star[0],mapastar_nombre))
print(getFromDict(star[0],mapastar_latitud))
print(getFromDict(star[0],mapastar_longitud))

In [ ]:
starbucks = []
for dic in star:
    st = {}
    st["name"] = getFromDict(dic,mapa_nombre)
    st["latitud"] = getFromDict(dic,m_latitud)
    st["longitud"] = getFromDict(dic,m_longitud)
    starbucks.append(st)

In [ ]:
starbucks[0]

In [ ]:
#Al comprobar los starbucks, me doy cuenta que existe sólo uno en Helsinki y me muestra otras categorías. 
df_st = pd.DataFrame(starbucks)
df_st.head()

In [ ]:
gdf4 = gpd.GeoDataFrame(df_st, geometry = gpd.points_from_xy(df_st.longitud,df_st.latitud))
gdf4.head()

In [ ]:
Map(Layer(gdf4, popup_hover = [popup_element("name","Starbucks")]))

### El Starbucks más cercano está a 500m: 
    
    Starbucks
    latitud: 60.167737
    longitud: 24.943369

## Quinta llamada: Restaurante Vegano

In [ ]:
params = {"client_id" : tok1,
              "client_secret" : tok2,
              "v": "20180323",
              "ll": f"60.17212, 24.94519",
              "query": "Vegan Restaurant",
              "limit": 20,
              "radius": 2000
         }
resp = requests.get(url=url1, params=params)
vegdata = json.loads(resp.text)


In [ ]:
decoding_data = vegdata.get("response")

In [ ]:
decoded = decoding_data.get("groups")[0]

In [ ]:
vegg = decoded.get("items")

In [ ]:
mapavegg_nombre = ["venue","name"]
mapavegg_latitud = ["venue","location","lat"]
mapavegg_longitud = ["venue","location","lng"]

In [ ]:
print(getFromDict(vegg[0],mapavegg_nombre))
print(getFromDict(vegg[0],mapavegg_latitud))
print(getFromDict(vegg[0],mapavegg_longitud))

In [ ]:
vegano = []
for dic in vegg:
    vg = {}
    vg["name"] = getFromDict(dic,mapavegg_nombre)
    vg["latitud"] = getFromDict(dic,mapavegg_latitud)
    vg["longitud"] = getFromDict(dic,mapavegg_longitud)
    vegano.append(vg)

In [ ]:
vegano[0]

In [ ]:
df_vg = pd.DataFrame(vegano)
df_vg.head()

In [ ]:
gdf5 = gpd.GeoDataFrame(df_vg, geometry = gpd.points_from_xy(df_vg.longitud,df_vg.latitud))
gdf5.head()

In [ ]:
Map(Layer(gdf5, popup_hover = [popup_element("name","Venga Restaurant")]))

### ¿Cuál es el restaurante vegano más cercano a mi empresa?

In [ ]:
gdf5['geometry']=gdf5['geometry'].apply(lambda x:shapely.geometry.mapping(x))
gdf5

In [ ]:
db = client.ironhack
collection = db.veganos
collection.create_index ([("geometry", GEOSPHERE)])

In [ ]:
data = gdf5.to_dict(orient='records')
collection.insert_many(data)

In [ ]:
data[:5]

In [ ]:
cercano5 = db.veganos.find({"geometry":{"$near":{"type":"Point", "coordinates":[60.171821,24.945167]}}})
list(cercano5)[:5]

### El Restaurante vegano más cercano está a 1.5km: 
    
    Pontus
    latitud: 60.1587920757416
    longitud: 24.94628194859813